aggregate_root/spec/snapshot_repository_spec.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+module AggregateRoot
+  class ReadingStatsRepository < SimpleDelegator
+    def initialize(repository)
+      super(repository)
+      @records_read = 0
+    end
+
+    attr_reader :records_read
+
+    def reset_records_read
+      @records_read = 0
+    end
+
+    def read(spec)
+      @records_read += count(spec)
+      super
+    end
+  end
+
+  RSpec.describe SnapshotRepository do
+    let(:event_store) do
+      RubyEventStore::Client.new(
+        repository: reporting_repository,
+        mapper: RubyEventStore::Mappers::NullMapper.new
+      )
+    end
+    let(:reporting_repository) { ReadingStatsRepository.new(RubyEventStore::InMemoryRepository.new) }
+    let(:uuid) { SecureRandom.uuid }
+    let(:stream_name) { "Order$#{uuid}" }
+    let(:repository) { AggregateRoot::SnapshotRepository.new(event_store) }
+    let(:order_created) { Orders::Events::OrderCreated.new }
+    let(:order_canceled) { Orders::Events::OrderCanceled.new }
+    let(:order_expired) { Orders::Events::OrderExpired.new }
+
+    let(:order_klass) do
+      class Order
+        include AggregateRoot
+
+        def initialize(uuid)
+          @status = :draft
+          @uuid = uuid
+        end
+
+        attr_accessor :status
+
+        def apply_order_created(_)
+          @status = :created
+        end
+
+        def apply_order_expired(_)
+          @status = :expired
+        end
+
+        def apply_order_canceled(_)
+          @status = :canceled
+        end
+      end
+
+      Order
+    end
+
+    specify 'initialization' do
+      expect { AggregateRoot::SnapshotRepository.new(event_store, 0) }
+        .to raise_error(ArgumentError, "interval must be greater than 0")
+    end
+
+    specify 'storing snapshot' do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store)
+      allow(event_store).to receive(:publish)
+
+      order.apply(order_created)
+      repository.store(order, stream_name)
+      expect(event_store).to have_received(:publish).with(
+        [order_created],
+        stream_name: stream_name,
+        expected_version: -1
+      )
+      expect_snapshot(order_created.event_id, 0, Marshal.dump(order))
+    end
+
+    specify 'storing snapshot with given interval' do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store, 2)
+      allow(event_store).to receive(:publish)
+
+      order.apply(order_created)
+      repository.store(order, stream_name)
+      expect_no_snapshot
+
+      order.apply(order_expired)
+      repository.store(order, stream_name)
+      expect_snapshot(order_expired.event_id, 1, Marshal.dump(order))
+    end
+
+    specify "restoring snapshot" do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store)
+
+      apply_and_publish(order, order_created, order_canceled)
```

Comment:
Could we lean more on aggregate here over using apply+publish helper? Perhaps we'd need to adjust snapshot interval but it would be less brittle and closer to how people use it.

```
order.create
order.cancel
repository.store(order)
```